internal/storage/storage.go

```diff
@@ -0,0 +1,60 @@
+package storage
+
+import (
+	"time"
+)
+
+// Backend is an interface for defining functions which all storage providers support.
+type Backend interface {
+	Copy(file string) error
+	Prune(deadline time.Time, pruningPrefix string) (*PruneStats, error)
+	GetName() string
+}
+
+// StorageBackend is a generic type of storage. Everything here are common properties of all storage types.
+type StorageBackend struct {
+	Backend
+	Name            string
+	DestinationPath string
+	RetentionDays   int
+	Log             LogFuncDef
+}
+
+type LogType string
+
+const (
+	INFO    LogType = "INFO"
+	WARNING LogType = "WARNING"
+	ERROR   LogType = "ERROR"
+)
+
+type LogFuncDef func(logType LogType, context string, msg string, params ...interface{}) error
+
+// PruneStats is a wrapper struct for returning stats after pruning
+type PruneStats struct {
+	Total  uint
+	Pruned uint
+}
+
+// DoPrune holds general control flow that applies to any kind of storage.
+// Callers can pass in a thunk that performs the actual deletion of files.
+func (stg *StorageBackend) DoPrune(lenMatches, lenCandidates int, description string, doRemoveFiles func() error) error {
```

Comment:
For the sake of consistency I'd prefer if we used single letter receiver names throughout, i.e. this would become `s *StorageBackend`. If you're afraid of collisions with `s *script` we could also use `b` here?

This also applies to the packages themselves.